### ISSUE TITLE
.transaction()

### ISSUE THREAD

**Issue by mikermcneil** [balderdashy/sails#6394]
@irlnathan @ponzicoder @particlebanana
Re our recent conversations about this subject

> **See bottom of this issue for proposed usage examples**
## tldr;

0.5) Add transaction support for uniqueness constraints (already done probably, since we're sending that down as part of the schema)

1) Add support for same-adapter transactions, using built-in stuff for the database  (e.g. mysql transactions).  The API looks like `User.transaction().find(....)`, `instance.rollback(cb)` and `instance.commit(cb)`.  If either of those fail, we're in deep trouble.  See the very bottom of this issue.

2) Add transaction support (using our new transaction API described above) on top of:
    + **aggregate** (`createEach()`, plural usage of `update()`, plural usage of `destroy()`)
    + and **composite** queries (`findOrCreate()`, and `findOrCreateEach()`)

3) Add `increment()` and `decrement()`, which are just shortcuts on top of update.  Only thing is they need to be transactional.  Using the same new functionality, of course.

4) Figure out how to handle commit() and rollback() errors.  This may already be worked out at the adapter level.

5) Add support for `instance.attempts()`, which will reattempt a waterline call _n_ times until it works (with incremental back-off).  We can probably leverage async for this.

6) Figure out how to do cross-adapter transactions in a user-friendly way.  This is explored in a NOT-user friendly way at the bottom of this issue.
## Motivations:

Ensuring isolated access to complex units of business logic, beyond what is available in a single database (since these logics may bridge multiple adapters)

From http://stackoverflow.com/a/14044329/486547:

> We started with full-on CRUD mutices, but soon realized that was pretty hard. Better to leave that to the database. But sometimes you don't want to-- because you want to be able to switch databases and keep your code agnostic. So then we simplified down to the next easiest piece-- named transactions.

Enabling this sort of cross-adapter behavior is crucial for allowing flexible usage of different services, and avoiding vendor lock-in (e.g. you could use the transactional storage quota support in S3, but what if you want to switch cloud storage providers?  Same thing could be said of the transaction support in Oracle.)
#### Problem (Signup Form)

> Keep in mind that this same scenario could exist if User A was making both requests- the trouble is the concurrency, not that the two users involved are coming from different sessions, _et al._
>   1) User A looks to see if any users exist with username "bronyboy47"
>   2) User B looks to see if any users exist with username "bronyboy47"
>   3) User A receives a response from Waterline and sees that no users yet exist with that username, so she calls `create()`.
>   4) User B receives a response from Waterline and sees that no users yet exist with that username, so she calls `create()`.
>   5) Two users with username "bronyboy47" are created.
#### Solution

Circumvent the need for transactions by using the unique constraint in your adapter.

We sidestep this issue by using the `unique` validation rule (as long as the adapter natively supports the `unique` validation-- otherwise we're screwed)  But in general, most production databases should support this, so who cares?  Get a new database, right?  What about `sails-disk` during development?  It would be pretty annoying to not be able to test your logic without hooking up your real production database!! Luckily, sails-disk uses an in-memory representation of your data model for simplicity, so all of its queries are transactional out of the box, since any requests against it are processed FIFO (which is a bottleneck-- but again, this is not a production situation!)  Good enough for me- let's stick to the low-hanging fruit.
### Increment/Decrement

This one's tougher.
#### Problem (Blackjack)

In the following example, a user is able to trick a Blackjack game.

  1) User A hits (i)
  2) User A hits (ii)
  3) User B hits (I)
  4) User B hits (II)
  5) User B's first request (I) is processed-- he gets dealt an "8", making the total value of his hand = 8.
  6) User A's first request (i) is processed-- she gets dealt a "9", making the total value of her hand = 9.
  7) User A hits (iii)
  7) User B hits (III)
  8) User B's second request (II) is processed-- he gets dealt an "8", making the total value of his hand = 16.
  9) User B's third request (III) is processed-- he gets dealt an "8", making the total value of his hand = 24.  User B busts. User A wins!  But then....
  10) User A's second request (ii) is processed-- she gets dealt a "9", making the total value of her hand = 18.
  11) User A's third request (iii) is processed-- she gets dealt a "9", making the total value of her hand = 27.  This is greater than 21 and she busts.  But wait.. she already won?
#### Solution

Use Hand's adapter's built-in support for transactions (TODO), plus Waterline's API for performing transactional increment/decrement (TODO).

``` javascript
var valueOfNewCard = Math.floor( Math.random() * 10 ) + 1; 
Hand.increment(currentHandId, {
  total: valueOfNewCard
}).done(cb);
```

 But let's dig in a bit deeper and use the raw transaction primitives:

Instead of `Hand`, use `Hand.transaction()`.  Instance models returned in the callback will also share access to the transaction.

Transactions are not technically written to the database until they're "committed".  However, they may also cause some or all of the database to lock, so it's important that, even in the event of failure, they are released (SQL calls this "rolled back").

Notice that any logic outside of Hand's adapter will need to be "transactional-ized" manually.  Within the database though, any updates, creates, or destroys will be automatically rolled back when the transaction releases the lock, either by calling "rollback" or "commit".

``` javascript
// Let's assume cb() is our exit

var dealtCard = Math.floor( Math.random() * 10 ) + 1; 
Hand
.transaction()
.findOne(3)
.done(function (err, hand) {
  if (hand.total > 21) {
    return hand.rollback(cb);
  }

  hand.total += dealtCard;
  hand.save().commit(cb);
})
```

When possible (i.e. adapter supports transactions natively, and all models involved belong to the same adapter), we should use the transaction support built into that adapter.  IMO, this is where Waterline's role ends-- single adapter transactions that have native support.  This functionality should be mixin-able in the adapter.  `findOrCreate()` should use built-in transactions automatically if they exist.

If rollback or commit fails, we're in a world of trouble.  See my comment below.
## Conclusion

To protect an atomic unit of business logic-- where logic which may touch more than one model, or **even more than one asynchronous service** (i.e. multiple adapters), we have a more complex problem, where we might want to get involved at the Sails level (i.e. locking middleware).  BUT!  As far as I can tell right now, multi-adapter transaction problems can always be simplified into a single-adapter problem.  I could quite easily be wrong though.  I'd love to get your ideas.

More on this below.
